TESTS/mbed_drivers/timerevent/main.cpp

```diff
@@ -0,0 +1,352 @@
+uint32_t mock_ticker_read_inc(void) {
+    static uint32_t cnt;
+    return ++cnt;
+}
+uint32_t mock_ticker_read_const(void) {
+    return MOCK_TICKER_CNT;
+}
+
+const ticker_interface_t mock_ticker_interface = {
+    .init = mock_ticker_fun,
+    .read = mock_ticker_read_inc,
+    .disable_interrupt = mock_ticker_fun,
+    .clear_interrupt = mock_ticker_fun,
+    .set_interrupt = mock_ticker_set_interrupt,
+    .fire_interrupt = mock_ticker_fun
+};
+
+/** Template for tests: ticker_queue handling
+ *
+ * Test ticker_queue after @a insert call
+ * Given an instance of @a TimerEvent subclass initialized with
+ *     an empty @a ticker_queue structure
+ * When a tiestamp is set with @a insert() call
+ * Then @a ticker_queue is set properly
+ * When the timestamp is removed
+ * Then @a ticker_queue is set properly
+ *
+ * Test ticker_queue after @a insert_absolute call
+ * Given an instance of @a TimerEvent subclass initialized with
+ *     an empty @a ticker_queue structure
+ * When a tiestamp is set with @a insert_absolute() call
+ * Then @a ticker_queue is set properly
+ * When the timestamp is removed
+ * Then @a ticker_queue is set properly
+ */
+template<typename T>
+void test_ticker_queue(void) {
+    ticker_event_queue_t ticker_queue = { 0 };
+    ticker_data_t ticker_data = {
+        &mock_ticker_interface,
+        &ticker_queue
+    };
+    us_timestamp_t prev_ts = 0;
+
+    T tte(&ticker_data);
+    TEST_ASSERT_TRUE(ticker_queue.initialized);
+    TEST_ASSERT_EQUAL_PTR(&TimerEvent::irq, ticker_queue.event_handler);
+    TEST_ASSERT(ticker_queue.present_time > prev_ts);
+    TEST_ASSERT_NULL(ticker_queue.head);
+    prev_ts = ticker_queue.present_time;
+
+    tte.set_timestamp(TEST_DELAY_US);
+    TEST_ASSERT_TRUE(ticker_queue.initialized);
+    TEST_ASSERT_EQUAL_PTR(&TimerEvent::irq, ticker_queue.event_handler);
+    TEST_ASSERT(ticker_queue.present_time > prev_ts);
+    TEST_ASSERT_EQUAL_PTR(tte.get_event_addr(), ticker_queue.head);
+    TEST_ASSERT_EQUAL(TEST_DELAY_US, ticker_queue.head->timestamp);
+    TEST_ASSERT_EQUAL_PTR(&tte, ticker_queue.head->id);
+    TEST_ASSERT_NULL(ticker_queue.head->next);
+    prev_ts = ticker_queue.present_time;
+
+    tte.remove();
+    TEST_ASSERT_TRUE(ticker_queue.initialized);
+    TEST_ASSERT_EQUAL_PTR(&TimerEvent::irq, ticker_queue.event_handler);
+    TEST_ASSERT(ticker_queue.present_time > prev_ts);
+    TEST_ASSERT_NULL(ticker_queue.head);
+}
+
+/** Template for tests: insert, insert_absolute
+ *
+ * Test insert
+ * Given an instance of @a TimerEvent subclass
+ * When a tiestamp is set with @a insert()
+ *     and given time elapses
+ * Then an event handler is called
+ *
+ * Test insert_absolute
+ * Given an instance of @a TimerEvent subclass
+ * When a tiestamp is set with @a insert_absolute()
+ *     and given time elapses
+ * Then an event handler is called
+ */
+template<typename T>
+void test_insert(void) {
+    T tte;
+    tte.sem_wait(0);
```

Comment:
Why are we decrementing semaphore here? You can just initialise the semaphore with count of 0.